I737 update docs for v9.9

clevengr opened the pull request:
### Description of what the PR does

Updates to the WhatsNew, Judges Guide, and Admin Guide `.docx` and `.pdf` files for V9.9.  
Sections which have been updated include:

**What's New**
- Added section for V9.9.0, pages 2 and 3.

**Judge's Guide**
- Updated title page date and version number.

**Admin Guide**
- Updated title page date and version number
- Corrected page numbers (so that the Table of Contents pages are number i,ii... etc. and the actual text starts at page number 1).
- Updated section "6.2.2 Account Names and Passwords":
    - Replaced screenshots with updated versions
    - Added descriptions of the Alias and Scoring Adjustment account configuration fields.
    - Added a brief explanation of "joe passwords", and a reference to the section on loading password (account) info.
- Updated section "6.2.3 Loading Account Data"
    - Added a description of the new "automatic loading" of "accounts_load.tsv" info.
- Updated Section "6.3.1 Defining a Problem" 
    - Replaced screenshots with updated versions.
    - Updated description of "Problem Letter" handling (documented that it's now shown on the GUI)
    - Added a paragraph explaining what an "interactive problem" is.
- Updated Section "6.3.2 Multiple Test Data Files"
    - Replaced screenshots with updated versions.
    - Documented the "Also load sample data" loading function.
- Updated screenshot in Section 6.3.3
- Updated Section "6.4.1 Defining a Language"
    - Updated the Languages tab screenshot
    - Updated the Add New Language dialog screenshot
    - Added documentation on the "Hide Language" and "ID/code" fields for a language.
    - Removed redundant screenshot.
- Updated Section "6.4.2 Command Parameter Substitutions"
    - Added documentation on the "{:package}" substitution variable for Java.
- Updated Section "6.4.3 Language Definition Examples"
    - Added an explanation of the use of "{:package}"
    - Updated the "Example" Java screenshot.
 - Regenerated level 1, 2, and 3 section headers for the entire document (because something glitched in MS Word during my edits and it was either regen the section headers or recreate all my changes...)
- Updated the "Team Settings" section of Section 6.7 Options (page 59) to include documentation on setting
   the team-scoreboard-display-format-string via YAML.  (See GitHub issue https://github.com/pc2ccs/pc2v9/issues/291).
- Added a note that "double-clicking" an account is another way to invoke the Edit Account dialog.
- Moved the information about automatic loading of accounts_load.tsv files from a footnote to the main text body.
- Added a note to section 6.3.2 paragraph 3 pointing out that external data on judge's machine must be IN THE SAME
   LOCATION IN THE FILE SYSTEM  on each machine.
- Added "ensuresuffix" to the table of language-related substitution variables in 6.4.3.
- Fixed a formatting problem in a footnote on p. 19.
- Added an additional footnote to the first paragraph in Appendix E providing additional details about where to look to find information interactive validators.
- Minor wording, margin, and section header formatting changes in Appendix U
- Added Section "4.4 Comparing Final Results Files" explaining the "Results Compare" tab to Appendix R (Shadowing).
- Added "Appendix V - The Judging Utilities Pane".
- Regenerated the Table of Contents.

### Issue which the PR addresses

Fixes #737 .    Also Fixes #291.


### Environment in which the PR was developed (OS,IDE, Java version, etc.)

Windows 11, Eclipse Version: 2023-03 (4.27.0). Java 1.8.0_371

### Precise steps for _testing_ the PR

- Read the Admin Guide sections identified as "updated" (above); verify that they provide accurate and adequate descriptions of their topics.
- Verify that section numbers throughout the document are consistent and that the Table of Contents contains correct page numbers.

johnbrvc commented:
In the Admin Guide, the page numbers in the _Table of Contest_ don't seem to reflect reality. ex.  For the first one, **6.2.2 Account Names and Passwords**, the TOC says it is on page 29.  The problem is, the title page and the roman numeral page numbers on the TOC are counted in the page count, yet the real pages start at page 5 (title + 4 pages of TOC).  Page 29 is really page 34  (title + 4 TOC pages + 29).

johnbrvc commented:
In 6.2.2, this may be been there before, but as a shortcut, you can "double-click" on an account to edit it directly instead of selecting it then pressing Edit.

johnbrvc commented:
6.2.3 - I think the "footnote" make it confusing.  The sentence with the footnote talks about using the Load button.  The footnote talks about automatically loading a specific file (accounts_load.tsv).  Why not just move this to the main text (same paragraph as where the footnote is used).

johnbrvc commented:
6.3.2 - I might add to paragraph 3: However, it requires the contest administrator to ensure that all the data files for the problem are copied onto all Judge’s machines **in the same location in the filesystem**.  (or some similar wording.)

johnbrvc commented:
6.4.2 - do we need to discuss #727  (where we added` {:ensuresuffix=_xxx_}` )

clevengr commented:
> In the Admin Guide, the page numbers in the _Table of Contest_ don't seem to reflect reality. ex. For the first one, **6.2.2 Account Names and Passwords**, the TOC says it is on page 29. The problem is, the title page and the roman numeral page numbers on the TOC are counted in the page count, yet the real pages start at page 5 (title + 4 pages of TOC). Page 29 is really page 34 (title + 4 TOC pages + 29).

Are you sure you're looking at the current PDF from my branch?   I did have such a problem with an earlier version, but the TOC in the PDF I'm looking at says that section 6.2.2 is on page 29, and indeed that's the number on the page in the document:

![image](https://github.com/pc2ccs/pc2v9/assets/9256691/1e08a1f0-9151-4b66-ab35-aad19da70a3b)

![image](https://github.com/pc2ccs/pc2v9/assets/9256691/515d9cc1-bc57-482a-875e-40310d4e5840)

johnbrvc commented:
> > In the Admin Guide, the page numbers in the _Table of Contest_ don't seem to reflect reality. ex. For the first one, **6.2.2 Account Names and Passwords**, the TOC says it is on page 29. The problem is, the title page and the roman numeral page numbers on the TOC are counted in the page count, yet the real pages start at page 5 (title + 4 pages of TOC). Page 29 is really page 34 (title + 4 TOC pages + 29).
> 
> Are you sure you're looking at the current PDF from my branch? I did have such a problem with an earlier version, but the TOC in the PDF I'm looking at says that section 6.2.2 is on page 29, and indeed that's the number on the page in the document:
> 
> ![image](https://user-images.githubusercontent.com/9256691/284118583-1e08a1f0-9151-4b66-ab35-aad19da70a3b.png)
> 
> ![image](https://user-images.githubusercontent.com/9256691/284118712-515d9cc1-bc57-482a-875e-40310d4e5840.png)

Yes.  The "page number" is right, but if you go to the top and key in the page number, it brings you to the wrong page.  
![image](https://github.com/pc2ccs/pc2v9/assets/5657363/dc1e6ffb-70f2-4677-803b-13be87613e8e)

clevengr commented:
@johnbrvc   \<rant\> This GitHub inability to keep a separate thread for a given comment is sure limiting... \</rant\>

Anyway, with regard to the page numbers being "wrong"... I _think_ what you're saying is that when you view the PDF in (some) PDF Viewer (possibly your browser?) and you tell the PDF Viewer to "go to page x", it's going to a page which is actually 5 away from x, due to the 1 title page + 4 TOC pages... is that correct?

If so, then doesn't that just mean that the PDF Viewer doesn't understand the page numbers in the footer of a Word document?  I'm not sure there's even a way to arrange for that to happen... is there?

One "fix" is to remove the "i, ii, ..." numbering in the TOC -- but I don't think that'll ignore the title page.  Anyway, I personally don't particularly care for the "style" where pages are numbered from the literal first piece of paper.  I'd much rather have front-matter identified with its own separate numbering scheme (like the traditional "i, ii, iii...")  

Do you have a suggestion for how to deal with this particular issue?

johnbrvc commented on doc/pc2v9AdminGuide.pdf:
I am testing this feature, but also commenting on @clevengr's comment about the page numbering issue being related to the PDF Viewer.  This is true, and there does not appear to be a valid work-around or solution for this.

clevengr commented:
> In 6.2.2, this may be been there before, but as a shortcut, you can "double-click" on an account to edit it directly instead of selecting it then pressing Edit.

Agreed; I've added a parenthetical remark to that effect (1st paragraph, page 30).  If you agree this fixes the particular issue, please mark this comment "resolved" (or else give me a more detailed suggestion).

clevengr commented:
> 6.3.2 - I might add to paragraph 3: However, it requires the contest administrator to ensure that all the data files for the problem are copied onto all Judge’s machines **in the same location in the filesystem**. (or some similar wording.)

Agreed; I've added the additional wording.

clevengr commented:
> 6.4.2 - do we need to discuss #727 (where we added` {:ensuresuffix=_xxx_}` ) ? Actually, the https://github.com/pc2ccs/pc2v9/wiki/Substitution-Variables should probably be updated to include that and the other new substitution variables like sandboxprogramname, sandboxcommandline, testcase, and a couple of other.

I've added `ensuresuffix` to the table in 6.4.2.  However, I don't want to move in the direction of trying to explain all the details in the Admin Guide; keeping it up to date is already way too cumbersome.  I agree that what we really need to do is update the wiki page with more details and with the missing substitution variables.  Perhaps a minimal starting point would be to add a paragraph, similar to the one just below the table in https://github.com/pc2ccs/pc2v9/wiki/YAML-Contest-Configuration#yaml-keywords, at least giving the names of variables which "exist but are not yet fully documented"?

clevengr commented:
I've just pushed a PR update which includes new copies of the Admin .docx and .pdf files containing updates per @johnbrvc 's suggestions.   Still working on updates for the remaining (missing) V9.9 topics.

clevengr commented:
> > 6.4.2 - do we need to discuss #727 (where we added` {:ensuresuffix=_xxx_}` ) ? Actually, the https://github.com/pc2ccs/pc2v9/wiki/Substitution-Variables should probably be updated to include that and the other new substitution variables like sandboxprogramname, sandboxcommandline, testcase, and a couple of other.
> 
> I've added `ensuresuffix` to the table in 6.4.2. However, I don't want to move in the direction of trying to explain all the details in the Admin Guide; keeping it up to date is already way too cumbersome. I agree that what we really need to do is update the wiki page with more details and with the missing substitution variables. Perhaps a minimal starting point would be to add a paragraph, similar to the one just below the table in https://github.com/pc2ccs/pc2v9/wiki/YAML-Contest-Configuration#yaml-keywords, at least giving the names of variables which "exist but are not yet fully documented"?

I've updated the Wiki page by adding a paragraph at the bottom listing some known (but undocumented) substitution variables.  At some point we need to add accurate documentation for those into the tables on the wiki page.

johnbrvc reviewed (approved):
Looks good.  Let me know if want me to check through it again if you make more changes.

troy2914 commented on doc/WhatsNew.pdf:
I think instead of saying "Improved compliance with CLICS Standards (see https://ccs-specs.icpc.io/), including"
you should have indicated which spec version where were going for 2020-03

clevengr replied:
The problem is that not all the updates (changes) relate to exactly the same standard (some of the changes were for 2020-03 while others were for 2022-07).  I've added explicit mentions of, and links to, _both_ standards.  Hopefully this solves the problem.

troy2914 commented on doc/WhatsNew.pdf:
"Upgrades to the WTI (Web Team Interface) bringing it up to more recent versions of Angular and Node."

i thought that was deferred to 9.10?

clevengr replied:
You're right; I've deleted that item (we'll have to try to remember to add it back when we generate a 9.10 WhatsNew...)

Although... GitHub is now forcing the use of Node 12, so THAT's what we're building into new distributions...

troy2914 commented on doc/pc2v9JudgesGuide.pdf:
page 3 has a footnote, that does not make sense.  if you are are doing bin/pc2judge having the PATH set to include the bin directory you would just run pc2judge.

clevengr replied:
Agreed.  I updated both the text and the footnote so that they are consistent with each other.

troy2914 commented on doc/pc2v9AdminGuide.pdf:
"The Run Timeout Limit value (shown as 10 seconds in the sample screen above but settable to any positive integer number of seconds when the problem is defined) is enforced by PC2"

seems you should mention that it is not enforced when running on the team machine.

clevengr replied:
Agreed; I've added a note to that effect.

troy2914 commented on doc/pc2v9AdminGuide.pdf:
"Contest problems in PC2 are global"
but filterable by groups.

clevengr replied:
Agreed; this is unclear.  In fact, that whole paragraph is a leftover from before we even had "groups".  I've updated it to reflect the existence of groups.

johnbrvc replied:
Hm. I'm not in love with the new wording.  Maybe: There is a "filter mechanism" that can be applied to control which problems are visible to teams belonging to specific groups (see the Groups...   Maybe "s/belonging to/that are members of/" ? It's probably OK the way it is, but it just reads a little clumsy (IMHO).

clevengr replied:
Ok, I've pushed an update that uses the new suggested wording, which I agree is better.

troy2914 commented on doc/pc2v9AdminGuide.pdf:
6.3.2  multiple test data files

it should be noted while loading into PC2 is the default only for the UI,  when doing an import the default is keep then external.

clevengr replied:
Agreed; I've reworded the paragraph and added a statement to that effect.

troy2914 reviewed (commented):
i left some comments on things that should probably be considered to update.

clevengr commented:
I have pushed new versions of the WhatsNew, Judge's Guide, and Admin Guide docs (Word source and corresponding PDFs) with updates that (hopefully) address the review comments @troy2914 made.  Waiting now for further reviews and/or another approval so this can be merged.

troy2914 commented on doc/WhatsNew.pdf:
That link for issues or is pr's does not work for me.   I am in the process of adding milestones, then we should use
a query like  https://github.com/pc2ccs/pc2v9/issues?q=is%3Aissue+milestone%3A9.9.0

clevengr replied:
@troy2914 Are you saying that you're going to go through every issue that's been resolved since the release of V9.8.0 (Dec 22, 2022) and tag them all with "Milestone 9.9.0"?

clevengr replied:
BTW:  yes, the link in the WhatsNew.pdf works for me; it lists every that's been closed between 22 Dec 2022 and 22 Nov 2023 and has a PR linked to it.   (Obviously, the terminal date should probably be updated since we didn't make the 22 Nov 2023 target... but the link appears to me to work...)

troy2914 replied:
well that link did not work for me,  and yes I have marked all the PR and ISSUES related to those PR's as Mileston 9.9.0

clevengr replied:
I'm curious:  what _happens_ for you when you try that link?

In any case, if you've marked all the relevant PRs and Issues as "Milestone 9.9.0", I agree it makes sense to update the link to what you've suggested; I'll push a PR update for that.  (I do think it makes more sense to do it using Milestones anyway; it avoids the issue of the "date range" such as what was in the original query.)

BTW, has anyone else noticed that this conversation is suddenly effectively like a Slack thread (instead of being a sequence of over-and-over "Quote in Reply" things)?   I wonder if GitHub made a change, or if there is just something different about the way this "thread" was started?

clevengr replied:
Ok, I've pushed an updated WhatsNew that fixes the link.  Let me know if there are any other changes to any of the docs that anybody wants to suggest.  If not, somebody else please mark it "approved" so we can merge it.

troy2914 replied:
when I entered the full link it just comes up a empty list.

johnbrvc replied:
Both the previous and the new link seem to work fine for me.

troy2914 reviewed (approved):
Seems good to me.
